File: cdm-sources/src/cdm/product/asset/calculation/functions/GetQuantityScheduleStepValues.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.base.math.NonNegativeQuantitySchedule import NonNegativeQuantitySchedule

__all__ = ['GetQuantityScheduleStepValues']


@replaceable
def GetQuantityScheduleStepValues(schedule: NonNegativeQuantitySchedule, periodStartDate: datetime.date) -> Decimal:
    """
    Find all schedule step values whose stepDate is before or equal to the supplied periodStartDate.  Returns a list of step values starting from the initial quantity value, to the last step value before the periodStartDate.
    
    Parameters 
    ----------
    schedule : NonNegativeQuantitySchedule
    The quantity schedule being looked up from.
    
    periodStartDate : date
    The date for which the quantity is required.
    
    Returns
    -------
    stepValues : number
    
    """
    self = inspect.currentframe()
    
    
    stepValues = rosetta_resolve_attr(rosetta_resolve_attr(self, "schedule"), "value")
    stepValues.add_rosetta_attr(self, (lambda item: map(lambda item: rosetta_resolve_attr(self, "value"), item))(rosetta_filter(rosetta_resolve_attr(rosetta_resolve_attr(self, "schedule"), "datedValue"), lambda item: all_elements(rosetta_resolve_attr(self, "date"), "<=", rosetta_resolve_attr(self, "periodStartDate")))))
    
    
    return stepValues

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/product/asset/calculation/functions/GetQuantityScheduleStepValues.java
package cdm.product.asset.calculation.functions;

import cdm.base.math.DatedValue;
import cdm.base.math.NonNegativeQuantitySchedule;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperS;
import com.rosetta.model.lib.records.Date;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(GetQuantityScheduleStepValues.GetQuantityScheduleStepValuesDefault.class)
public abstract class GetQuantityScheduleStepValues implements RosettaFunction {

	/**
	* @param schedule The quantity schedule being looked up from.
	* @param periodStartDate The date for which the quantity is required.
	* @return stepValues 
	*/
	public List<BigDecimal> evaluate(NonNegativeQuantitySchedule schedule, Date periodStartDate) {
		List<BigDecimal> stepValues = doEvaluate(schedule, periodStartDate);
		
		return stepValues;
	}

	protected abstract List<BigDecimal> doEvaluate(NonNegativeQuantitySchedule schedule, Date periodStartDate);

	public static class GetQuantityScheduleStepValuesDefault extends GetQuantityScheduleStepValues {
		@Override
		protected List<BigDecimal> doEvaluate(NonNegativeQuantitySchedule schedule, Date periodStartDate) {
			List<BigDecimal> stepValues = new ArrayList<>();
			return assignOutput(stepValues, schedule, periodStartDate);
		}
		
		protected List<BigDecimal> assignOutput(List<BigDecimal> stepValues, NonNegativeQuantitySchedule schedule, Date periodStartDate) {
			stepValues.addAll(MapperS.of(schedule).<BigDecimal>map("getValue", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getValue()).getMulti());
			
			final MapperC<DatedValue> thenArg = MapperS.of(schedule).<DatedValue>mapC("getDatedValue", nonNegativeQuantitySchedule -> nonNegativeQuantitySchedule.getDatedValue())
				.filterItemNullSafe(item -> lessThanEquals(item.<Date>map("getDate", datedValue -> datedValue.getDate()), MapperS.of(periodStartDate), CardinalityOperator.All).get());
			stepValues.addAll(thenArg
				.mapItem(item -> item.<BigDecimal>map("getValue", datedValue -> datedValue.getValue())).getMulti());
			
			return stepValues;
		}
	}
}
